worlds/stardew_valley/logic.py

```diff
@@ -607,136 +685,260 @@ def has_house(self, upgrade_level: int) -> StardewRule:
+        farm_rule = self.can_reach_region(Region.farm) | self.can_reach_region(
+            Region.greenhouse) | self.has_island_farm()
+        return season_rule & farm_rule
+
+    def has_island_farm(self) -> StardewRule:
+        return self.can_reach_region(Region.island_south)
 
     def can_catch_fish(self, fish: FishItem) -> StardewRule:
         region_rule = self.can_reach_any_region(fish.locations)
         season_rule = self.has_any_season(fish.seasons)
-        difficulty_rule = self.can_fish(fish.difficulty)
         if fish.difficulty == -1:
             difficulty_rule = self.can_crab_pot()
+        else:
+            difficulty_rule = self.can_fish(fish.difficulty)
         return region_rule & season_rule & difficulty_rule
 
     def can_catch_every_fish(self) -> StardewRule:
-        rules = [self.has_skill_level("Fishing", 10), self.has_max_fishing_rod()]
+        rules = [self.has_skill_level(Skill.fishing, 10), self.has_max_fishing_rod()]
         for fish in all_fish:
+            if self.options[options.ExcludeGingerIsland] == options.ExcludeGingerIsland.option_true and \
+                    fish in island_fish:
+                continue
             rules.append(self.can_catch_fish(fish))
         return And(rules)
 
     def has_max_fishing_rod(self) -> StardewRule:
         if self.options[options.ToolProgression] == options.ToolProgression.option_progressive:
-            return self.received("Progressive Fishing Rod", 4)
+            return self.received(APTool.fishing_rod, 4)
         return self.can_get_fishing_xp()
 
-    def can_cook(self) -> StardewRule:
-        return self.has_house(1) or self.has_skill_level("Foraging", 9)
+    def can_cook(self, recipe: CookingRecipe = None) -> StardewRule:
+        cook_rule = self.has_house(1) | self.has_skill_level(Skill.foraging, 9)
+        if recipe is None:
+            return cook_rule
+
+        learn_rule = self.can_learn_recipe(recipe.source)
+        ingredients_rule = And([self.has(ingredient) for ingredient in recipe.ingredients])
+        number_ingredients = sum(recipe.ingredients[ingredient] for ingredient in recipe.ingredients)
+        time_rule = self.has_lived_months(number_ingredients)
+        return cook_rule & learn_rule & ingredients_rule & time_rule
+
+    def can_learn_recipe(self, source: RecipeSource) -> StardewRule:
+        if isinstance(source, StarterSource):
+            return True_()
+        if isinstance(source, ShopSource):
+            return self.can_spend_money_at(source.region, source.price)
+        if isinstance(source, SkillSource):
+            return self.has_skill_level(source.skill, source.level)
+        if isinstance(source, FriendshipSource):
+            return self.has_relationship(source.friend, source.hearts)
+        if isinstance(source, QueenOfSauceSource):
+            year_rule = self.has_year_two() if source.year == 2 else self.has_year_three()
```

Comment:
I'm unsure why ```self.has_year_three()``` is used here. Wouldn't this make Queen of Sauce recipes with ```source.year == 1``` take _longer_ than those with ```source.year == 2```?

Reply:
Honestly, this one is very "prototype". Basically, yes, year 1 recipes are expected after year 2 recipes. This is because I don't actually know when the player received the cooking channel.

For example, if they get it right as year 2 starts, then logic would expect all recipes to be learnable, while it's an extra year to loop around to the year 1 recipes.

So I decided to make them year 2 and 3, assuming that the channel can be received within that delay and then the gameplay will match the logic, mostly.

But I fully agree it's not great, and with chefsanity in 5.x.x, I plan on revisiting this and maybe changing how queen of sauce works client-side so I can write more accurate logic for it.